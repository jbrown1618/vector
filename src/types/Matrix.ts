import { LinearTransformation } from './LinearTransformation';
import { Vector, VectorData } from './Vector';
import { ScalarContainer } from './ScalarContainer';

export type MatrixData<ScalarType> = Array<VectorData<ScalarType>>;
export type MatrixEntryCallback<ScalarType> = (
  entry: ScalarType,
  rowIndex: number,
  columnIndex: number
) => void;

export interface Matrix<ScalarType>
  extends LinearTransformation<Vector<ScalarType>, Vector<ScalarType>>,
    ScalarContainer<ScalarType> {
  /**
   * @returns {MatrixData<ScalarType>}  The contents of the matrix as an array of arrays.
   */
  getData(): MatrixData<ScalarType>;

  /**
   * @returns {number}  The number of rows in the matrix
   */
  getNumberOfRows(): number;

  /**
   * @returns {number}  The number of columns in the matrix
   */
  getNumberOfColumns(): number;

  /**
   * @returns {Array<Vector<ScalarType>>} An array of vectors corresponding to the rows of the matrix
   */
  getRowVectors(): Array<Vector<ScalarType>>;

  /**
   * @param {number} rowIndex  The index for which to fetch the row
   * @returns {Vector<ScalarType>}  A vector corresponding to the row at index `rowIndex`
   */
  getRow(rowIndex: number): Vector<ScalarType>;

  /**
   * @returns {Array<Vector<ScalarType>>}  An array of vectors corresponding to the columns of the matrix
   */
  getColumnVectors(): Array<Vector<ScalarType>>;

  /**
   * @param {number} columnIndex  The index for which to fetch the column
   * @returns {Vector<ScalarType>}  A vector corresponding to the column at index `columnIndex`
   */
  getColumn(columnIndex: number): Vector<ScalarType>;

  /**
   * @param {number} rowIndex  The index of the row containing the entry
   * @param {number} columnIndex  The index of the column containing the entry
   * @returns {ScalarType}  The entry located at `(rowIndex, columnIndex)`
   */
  getEntry(rowIndex: number, columnIndex: number): ScalarType;

  /**
   * Returns a *new* matrix equal to the old one, except with the entry at
   * `(rowIndex, columnIndex)` replaced with `value`
   *
   * @param {number} rowIndex  The row containing the value to replace
   * @param {number} columnIndex  The column containing the value to replace
   * @param {ScalarType} value  The new value
   * @returns {Matrix<ScalarType>}  A new matrix
   */
  set(rowIndex: number, columnIndex: number, value: ScalarType): Matrix<ScalarType>;

  /**
   * Implements matrix multiplication
   * @param {Matrix<ScalarType>} other  The matrix by which to multiply
   * @returns {Matrix<ScalarType>}  The matrix product
   */
  multiply(other: Matrix<ScalarType>): Matrix<ScalarType>;

  /**
   * @returns {Matrix<ScalarType>}  The adjoint of the matrix
   */
  adjoint(): Matrix<ScalarType>;

  /**
   * @returns {Matrix<ScalarType>}  The transpose of the matrix
   */
  transpose(): Matrix<ScalarType>;

  /**
   * Implements matrix addition
   * @param {Matrix<ScalarType>} other  The matrix to add
   * @returns {Matrix<ScalarType>}  The matrix sum
   */
  add(other: Matrix<ScalarType>): Matrix<ScalarType>;

  /**
   * Implements multiplication of a matrix by a scalar
   * @param {ScalarType} scalar  The scalar by which to multiply
   * @returns {Matrix<ScalarType>}  The product
   */
  scalarMultiply(scalar: ScalarType): Matrix<ScalarType>;

  /**
   * @param {Matrix<ScalarType>} other
   * @returns {boolean}  true if `this` is equal to `other`
   */
  equals(other: Matrix<ScalarType>): boolean;

  /**
   * Executes the `callback` function for each entry in the matrix.
   * @param {MatrixEntryCallback<ScalarType>} callback
   */
  forEachEntry(callback: MatrixEntryCallback<ScalarType>): void;
}
