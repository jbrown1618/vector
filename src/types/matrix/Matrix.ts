import { ScalarOperations } from '../scalar/ScalarOperations';
import { Vector, VectorData } from '../vector/Vector';
import { VectorBuilder } from '../vector/VectorBuilder';
import { LinearTransformation } from './LinearTransformation';
import { MatrixBuilder } from './MatrixBuilder';

/**
 * @public
 */
export type MatrixData<S> = readonly VectorData<S>[];

/**
 * @public
 */
export type MatrixEntryCallback<S> = (entry: S, rowIndex: number, columnIndex: number) => void;

/**
 * @public
 */
export interface MatrixConstructor<S, V extends Vector<S>, M extends Matrix<S>> {
  new (data: MatrixData<S>): M;
  ops(): ScalarOperations<S>;
  builder(): MatrixBuilder<S, V, M>;
  vectorBuilder(): VectorBuilder<S, V>;
}

/**
 * @public
 */
export interface Matrix<S> extends LinearTransformation<Vector<S>, Vector<S>> {
  /**
   * Yields a `ScalarOperations` object which will allow consumers to work generically
   * with the scalars contained in the vector.
   */
  ops(): ScalarOperations<S>;

  /**
   * Yields a `MatrixBuilder` which will build new matrices of the same type
   */
  builder(): MatrixBuilder<S, Vector<S>, Matrix<S>>;

  /**
   * Yields a `VectorBuilder` which will build new vectors of a compatible type
   */
  vectorBuilder(): VectorBuilder<S, Vector<S>>;

  /**
   * @returns The contents of the matrix as an array of arrays.
   * @public
   */
  toArray(): S[][];

  /**
   * @returns The contents of the matrix as a nested map of rowIndex to columnIndex to nonzero value
   * @public
   */
  getSparseData(): Map<number, Map<number, S>>;

  /**
   * @returns The number of rows in the matrix
   * @public
   */
  getNumberOfRows(): number;

  /**
   * @returns The number of columns in the matrix
   * @public
   */
  getNumberOfColumns(): number;

  /**
   * @returns An array of vectors corresponding to the rows of the matrix
   * @public
   */
  getRowVectors(): Vector<S>[];

  /**
   * @param rowIndex - The index for which to fetch the row
   * @returns A vector corresponding to the row at index `rowIndex`
   * @public
   */
  getRow(rowIndex: number): Vector<S>;

  /**
   * @returns An array of vectors corresponding to the columns of the matrix
   * @public
   */
  getColumnVectors(): Vector<S>[];

  /**
   * @param columnIndex - The index for which to fetch the column
   * @returns A vector corresponding to the column at index `columnIndex`
   * @public
   */
  getColumn(columnIndex: number): Vector<S>;

  /**
   * @returns A vector containing the diagonal elements of the matrix
   * @public
   */
  getDiagonal(): Vector<S>;

  /**
   * @param rowIndex - The index of the row containing the entry
   * @param columnIndex - The index of the column containing the entry
   * @returns The entry located at `(rowIndex, columnIndex)`
   * @public
   */
  getEntry(rowIndex: number, columnIndex: number): S;

  /**
   * Returns a *new* matrix equal to the old one, except with the entry at
   * `(rowIndex, columnIndex)` replaced with `value`
   *
   * @param rowIndex - The row containing the value to replace
   * @param columnIndex - The column containing the value to replace
   * @param value - The new value
   * @returns A new matrix
   * @public
   */
  set(rowIndex: number, columnIndex: number, value: S): Matrix<S>;

  /**
   * Implements matrix multiplication
   * @param other - The matrix by which to multiply
   * @returns The matrix product
   * @public
   */
  multiply(other: Matrix<S>): Matrix<S>;

  /**
   * Applies the matrix as a linear transformation to the given vector.
   * Implements matrix-vector multiplication.
   * @param vector - the vector that should be transformed by the matrix
   * @returns The transformed vector
   * @public
   */
  apply(vector: Vector<S>): Vector<S>;

  /**
   * @returns The adjoint of the matrix
   * @public
   */
  adjoint(): Matrix<S>;

  /**
   * @returns The transpose of the matrix
   * @public
   */
  transpose(): Matrix<S>;

  /**
   * @returns The trace of the matrix
   * @public
   */
  trace(): S;

  /**
   * Implements matrix addition
   * @param other - The matrix to add
   * @returns The matrix sum
   * @public
   */
  add(other: Matrix<S>): Matrix<S>;

  /**
   * Implements multiplication of a matrix by a scalar
   * @param scalar - The scalar by which to multiply
   * @returns The product
   * @public
   */
  scalarMultiply(scalar: S): Matrix<S>;

  /**
   * @param other - The matrix against which to compare
   * @returns true if `this` is equal to `other`
   * @public
   */
  equals(other: Matrix<S>): boolean;

  /**
   * Executes the `callback` function for each entry in the matrix.
   * @param callback - The function to execute for each entry
   * @public
   */
  forEachEntry(callback: MatrixEntryCallback<S>): void;
}
