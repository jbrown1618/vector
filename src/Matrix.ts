import { LinearTransformation } from './LinearTransformation';
import { Vector, VectorData } from './Vector';
import { AbstractVector } from './AbstractVector';
import { MatrixBuilder } from './utilities/MatrixBuilder';
import { reducedRowEchelonForm } from './algorithms/RowEchelon';

export type MatrixData = Array<VectorData>;
export type MatrixEntryCallback = (entry: number, rowIndex: number, columnIndex: number) => void;

export class Matrix implements LinearTransformation<Vector, Vector>, AbstractVector<Matrix> {
  private readonly _data: MatrixData;

  private constructor(data: MatrixData) {
    if (data.length > 0) {
      const rowDimension = data[0].length;
      for (let i = 0; i < data.length; i++) {
        if (data[i].length != rowDimension) {
          throw Error('Non-rectangular data');
        }
      }

      if (rowDimension === 0) {
        data = [];
      }
    }
    this._data = data;
  }

  static fromData(data: MatrixData) {
    return new Matrix(data);
  }

  static fromColumnVectors(columns: Array<Vector>) {
    const data: MatrixData = [];

    if (columns.length === 0) {
      return Matrix.fromData(data);
    }

    const columnDimension = columns[0].getDimension();
    columns.forEach(columnVector => {
      if (columnVector.getDimension() != columnDimension) {
        throw new Error('Column vectors must all have the same dimension');
      }
    });

    // initialize rows
    for (let i = 0; i < columnDimension; i++) {
      data[i] = [];
    }

    columns.forEach((columnVector, columnIndex) => {
      columnVector.getData().forEach((value, rowIndex) => {
        data[rowIndex][columnIndex] = value;
      });
    });

    return Matrix.fromData(data);
  }

  static fromRowVectors(rows: Array<Vector>): Matrix {
    const data: MatrixData = [];

    if (rows.length === 0) {
      return Matrix.fromData(data);
    }

    const columnDimension = rows[0].getDimension();
    rows.forEach(columnVector => {
      if (columnVector.getDimension() != columnDimension) {
        throw new Error('Column vectors must all have the same dimension');
      }
    });

    return Matrix.fromData(rows.map(row => row.getData()));
  }

  public getData(): MatrixData {
    return this.getRowVectors().map(row => row.getData());
  }

  public getNumberOfRows(): number {
    return this.getRowVectors().length;
  }

  public getNumberOfColumns(): number {
    return this.getColumnVectors().length;
  }

  public getRowVectors(): Array<Vector> {
    return this._data.map(row => Vector.fromData(row));
  }

  public getRow(rowIndex: number): Vector {
    if (rowIndex > this.getNumberOfRows() - 1 || rowIndex < 0) {
      throw new Error('Index out of bounds');
    }
    return this.getRowVectors()[rowIndex];
  }

  public getColumnVectors(): Array<Vector> {
    return this.transpose().getRowVectors();
  }

  public getColumn(columnIndex: number): Vector {
    if (columnIndex > this.getNumberOfColumns() - 1 || columnIndex < 0) {
      throw new Error('Index out of bounds');
    }
    return this.getColumnVectors()[columnIndex];
  }

  public getEntry(rowIndex: number, columnIndex: number) {
    return this.getRow(rowIndex).getEntry(columnIndex);
  }

  public set(rowIndex: number, columnIndex: number, value: number) {
    const copy = this.getData();
    copy[rowIndex][columnIndex] = value;
    return Matrix.fromData(copy);
  }

  public getDimension(): number {
    return this.getNumberOfRows() * this.getNumberOfColumns();
  }

  public multiply(other: Matrix): Matrix {
    if (this.getNumberOfColumns() !== other.getNumberOfRows()) {
      throw new Error('Dimension mismatch');
    }

    return Matrix.fromData(
      this.getRowVectors().map(row =>
        other.getColumnVectors().map(column => row.innerProduct(column))
      )
    );
  }

  public transpose(): Matrix {
    return Matrix.fromColumnVectors(this.getRowVectors());
  }

  public isSquare(): boolean {
    return this.getNumberOfColumns() == this.getNumberOfRows();
  }

  public inverse(): Matrix {
    if (!this.isSquare()) {
      throw Error('Non-square matrix has no inverse');
    }

    const augmented = MatrixBuilder.augment(
      this,
      MatrixBuilder.identity(this.getNumberOfColumns())
    );
    const rref = reducedRowEchelonForm(augmented);
    return MatrixBuilder.slice(rref, 0, this.getNumberOfColumns());
  }

  public add(other: Matrix): Matrix {
    return Matrix.fromColumnVectors(
      this.getColumnVectors().map((column, columnIndex) => column.add(other.getColumn(columnIndex)))
    );
  }

  public apply(vector: Vector): Vector {
    const vectorAsColumnMatrix = Matrix.fromColumnVectors([vector]);
    return this.multiply(vectorAsColumnMatrix).getColumn(0);
  }

  public innerProduct(other: Matrix): number {
    return this.getRowVectors()
      .map((row, rowIndex) => row.innerProduct(other.getRow(rowIndex)))
      .reduce((sum, next) => sum + next);
  }

  public scalarMultiply(scalar: number): Matrix {
    return Matrix.fromColumnVectors(
      this.getColumnVectors().map(column => column.scalarMultiply(scalar))
    );
  }

  public approxEquals(other: Matrix): boolean {
    if (this.getNumberOfColumns() !== other.getNumberOfColumns()) {
      return false;
    }

    if (this.getNumberOfRows() !== other.getNumberOfRows()) {
      return false;
    }

    return this.getColumnVectors()
      .map((column, i) => column.approxEquals(other.getColumn(i)))
      .reduce((all, current) => all && current, true);
  }

  public forEachEntry(callback: MatrixEntryCallback) {
    this.getRowVectors().forEach((row, rowIndex) => {
      row.getData().forEach((entry, columnIndex) => {
        callback(entry, rowIndex, columnIndex);
      });
    });
  }
}
