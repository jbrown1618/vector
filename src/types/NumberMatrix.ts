import { ArrayMatrix } from './ArrayMatrix';
import { MatrixConstructor, MatrixData } from './Matrix';
import { NumberOperations } from './NumberOperations';
import { StaticImplements } from '../utilities/StaticImplements';
import { MatrixBuilder, NumberVector, VectorBuilder } from '..';

/**
 * A `Matrix` implemented as a 2-dimensional array of JS `number` primitives
 */
@StaticImplements<MatrixConstructor<number, NumberVector, NumberMatrix>>()
export class NumberMatrix extends ArrayMatrix<number> {
  static ops() {
    return new NumberOperations();
  }

  static builder(): MatrixBuilder<number, NumberVector, NumberMatrix> {
    return new MatrixBuilder(NumberMatrix);
  }

  static vectorBuilder(): VectorBuilder<number, NumberVector> {
    return new VectorBuilder(NumberVector);
  }

  ops() {
    return NumberMatrix.ops();
  }

  builder(): MatrixBuilder<number, NumberVector, NumberMatrix> {
    return NumberMatrix.builder();
  }

  vectorBuilder(): VectorBuilder<number, NumberVector> {
    return NumberMatrix.vectorBuilder();
  }

  constructor(data: MatrixData<number>) {
    super(data);
  }
}
