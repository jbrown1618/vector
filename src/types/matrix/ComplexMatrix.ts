import { StaticImplements } from '../../utilities/StaticImplements';
import { ComplexNumber } from '../scalar/ComplexNumber';
import { ComplexNumberOperations } from '../scalar/ComplexNumberOperations';
import { ComplexVector } from '../vector/ComplexVector';
import { VectorBuilder } from '../vector/VectorBuilder';
import { ArrayMatrix } from './ArrayMatrix';
import { MatrixConstructor, MatrixData } from './Matrix';
import { MatrixBuilder } from './MatrixBuilder';
import { ScalarOperations } from '../scalar/ScalarOperations';

/**
 * Implements `Matrix` as a 2-dimensional array of `ComplexNumber`s
 * @public
 */
@StaticImplements<MatrixConstructor<ComplexNumber, ComplexVector, ComplexMatrix>>()
export class ComplexMatrix extends ArrayMatrix<ComplexNumber> {
  public static ops(): ScalarOperations<ComplexNumber> {
    return new ComplexNumberOperations();
  }

  public static builder(): MatrixBuilder<ComplexNumber, ComplexVector, ComplexMatrix> {
    return new MatrixBuilder(ComplexMatrix);
  }

  public static vectorBuilder(): VectorBuilder<ComplexNumber, ComplexVector> {
    return new VectorBuilder(ComplexVector);
  }

  constructor(data: MatrixData<ComplexNumber>) {
    super(data);
  }

  /**
   * {@inheritDoc TODO}
   */
  public ops(): ScalarOperations<ComplexNumber> {
    return ComplexMatrix.ops();
  }

  /**
   * {@inheritDoc TODO}
   */
  public builder(): MatrixBuilder<ComplexNumber, ComplexVector, ComplexMatrix> {
    return ComplexMatrix.builder();
  }

  /**
   * {@inheritDoc TODO}
   */
  public vectorBuilder(): VectorBuilder<ComplexNumber, ComplexVector> {
    return ComplexMatrix.vectorBuilder();
  }
}
