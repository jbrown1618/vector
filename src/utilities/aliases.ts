import { Vector } from '@lib/types/vector/Vector';
import { Matrix } from '@lib/types/matrix/Matrix';
import { NumberVector } from '@lib/types/vector/NumberVector';
import { NumberMatrix } from '@lib/types/matrix/NumberMatrix';

const vb = NumberVector.builder();
const mb = NumberMatrix.builder();

/**
 * Creates a new {@link Vector} of numbers.  See {@link VectorBuilder.fromArray}
 * @public
 */
export function vec(data: number[]): Vector<number> {
  return vb.fromArray(data);
}

/**
 * Creates a new {@link Matrix} of numbers.  See {@link MatrixBuilder.fromArray}
 * @public
 */
export function mat(data: number[][]): Matrix<number> {
  return mb.fromArray(data);
}

/**
 * Creates a new identity matrix of size `size`.  See {@link MatrixBuilder.identity}
 * @public
 */
export function eye(size: number): Matrix<number> {
  return mb.identity(size);
}

/**
 * Creates a new vector of all 1s.  See {@link VectorBuilder.ones}
 * @public
 */
export function ones(entries: number): Vector<number>;
/**
 * Creates a new matrix of all 1s.  See {@link MatrixBuilder.ones}
 * @public
 */
export function ones(rows: number, columns: number): Matrix<number>;
export function ones(rows: number, columns?: number): Matrix<number> | Vector<number> {
  return columns === undefined ? vb.ones(rows) : mb.ones(rows, columns);
}

/**
 * Creates a new vector of all 0s.  See {@link VectorBuilder.zeros}
 * @public
 */
export function zeros(entries: number): Vector<number>;
/**
 * Creates a new matrix of all 0s.  See {@link MatrixBuilder.zeros}
 * @public
 */
export function zeros(rows: number, columns: number): Matrix<number>;
export function zeros(rows: number, columns?: number): Matrix<number> | Vector<number> {
  return columns === undefined ? vb.zeros(rows) : mb.zeros(rows, columns);
}

/**
 * Creates a new matrix with the specified entries on the diagonal.  See {@link MatrixBuilder.diagonal}
 * @public
 */
export function diag(elements: number[]): Matrix<number> {
  return mb.diagonal(vec(elements));
}
