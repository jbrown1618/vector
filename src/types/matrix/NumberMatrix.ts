import { StaticImplements } from '../../utilities/StaticImplements';
import { NumberOperations } from '../scalar/NumberOperations';
import { NumberVector } from '../vector/NumberVector';
import { VectorBuilder } from '../vector/VectorBuilder';
import { ArrayMatrix } from './ArrayMatrix';
import { MatrixConstructor, MatrixData } from './Matrix';
import { MatrixBuilder } from './MatrixBuilder';
import { ScalarOperations } from '../scalar/ScalarOperations';

/**
 * A `Matrix` implemented as a 2-dimensional array of JS `number` primitives
 * @public
 */
@StaticImplements<MatrixConstructor<number, NumberVector, NumberMatrix>>()
export class NumberMatrix extends ArrayMatrix<number> {
  public static ops(): ScalarOperations<number> {
    return new NumberOperations();
  }

  public static builder(): MatrixBuilder<number, NumberVector, NumberMatrix> {
    return new MatrixBuilder(NumberMatrix);
  }

  public static vectorBuilder(): VectorBuilder<number, NumberVector> {
    return new VectorBuilder(NumberVector);
  }

  constructor(data: MatrixData<number>) {
    super(data);
  }

  /**
   * {@inheritDoc ArrayMatrix.ops}
   */
  public ops(): ScalarOperations<number> {
    return NumberMatrix.ops();
  }

  /**
   * {@inheritDoc ArrayMatrix.builder}
   */
  public builder(): MatrixBuilder<number, NumberVector, NumberMatrix> {
    return NumberMatrix.builder();
  }

  /**
   * {@inheritDoc ArrayMatrix.vectorBuilder}
   */
  public vectorBuilder(): VectorBuilder<number, NumberVector> {
    return NumberMatrix.vectorBuilder();
  }
}
