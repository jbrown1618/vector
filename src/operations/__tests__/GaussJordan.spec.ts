import { mat, vec, eye } from '../../utilities/aliases';
import { Matrix } from '../../types/matrix/Matrix';
import { NumberMatrix } from '../../types/matrix/NumberMatrix';
import {
  inverse,
  reducedRowEchelonForm,
  rowEchelonForm,
  solveByGaussianElimination
} from '../GaussJordan';
import {
  SolutionType,
  UnderdeterminedSolution,
  UniqueSolution
} from '../../solvers/LinearSolution';
import { loadTestData } from '@test-utils/testData';

describe('GaussJordan', () => {
  describe('solveByGaussianElimination', () => {
    test('solves a system with a unique solution', () => {
      const A = mat([
        [0, 2, 1],
        [1, -2, -3],
        [-1, 1, 2]
      ]);
      const b = vec([-8, 0, 3]);
      const solution = solveByGaussianElimination(A, b);
      const expected = vec([-4, -5, 2]);

      expect(solution.solutionType).toEqual(SolutionType.UNIQUE);
      expect((solution as UniqueSolution<number>).solution).toStrictEqual(expected);
    });

    test('solves a random 20x20 system', () => {
      const A = loadTestData('random-20x20');
      const b = loadTestData('random-20x1').getColumn(0);

      const solution = solveByGaussianElimination(A, b);

      expect(solution).toMatchSnapshot();
      expect(solution.solutionType).toBe(SolutionType.UNIQUE);
      const x = (solution as UniqueSolution<number>).solution;
      expect(A.apply(x).equals(b)).toBe(true);
    });

    test('solves an underdetermined system', () => {
      const A = mat([
        [0, 2, 1],
        [1, -2, -3],
        [-3, 6, 9]
      ]);
      const b = vec([-4, 1, -3]);
      const solution = solveByGaussianElimination(A, b);
      const expected = vec([-1, -5 / 2, 1]);

      expect(solution.solutionType).toEqual(SolutionType.UNDERDETERMINED);
      expect((solution as UnderdeterminedSolution<number>).solution).toStrictEqual(expected);
    });

    test('determines when a system is overdetermined', () => {
      const A = mat([
        [1, -2, -6],
        [2, 4, 12],
        [1, -4, -12]
      ]);
      const b = vec([12, -17, 22]);
      const solution = solveByGaussianElimination(A, b);

      expect(solution.solutionType).toEqual(SolutionType.OVERDETERMINED);
    });
  });

  describe('rowEchelonForm', () => {
    test('row reduces a "wide" matrix', () => {
      const A = mat([
        [1, 2, 3],
        [4, 5, 6]
      ]);
      const aRef = rowEchelonForm(A);
      expect(aRef.toArray()).toStrictEqual([
        [4, 5, 6],
        [0, 0.75, 1.5]
      ]);
    });

    test('row reduces a "tall" matrix', () => {
      const A = mat([
        [0, 1],
        [0, 0],
        [5, 9]
      ]);
      const aRef = rowEchelonForm(A);
      const expected = mat([
        [5, 9],
        [0, 1],
        [0, 0]
      ]);
      expect(aRef.equals(expected)).toBe(true);
    });

    test('row reduces a matrix with non-independent rows', () => {
      const A = mat([
        [1, 2, 3],
        [1, 1, 1],
        [1, 1, 1]
      ]);
      const aRef = rowEchelonForm(A);
      expect(aRef).toStrictEqual(
        mat([
          [1, 2, 3],
          [0, -1, -2],
          [0, 0, 0]
        ])
      );

      // Matrix B has an unexpected row of zeros occur, necessitating an extra swap
      const B = mat([
        [1, 2, 2, 0],
        [-1, -2, -2, 0],
        [-3, 9, 9, 0]
      ]);
      const bRef = rowEchelonForm(B);
      expect(bRef).toStrictEqual(
        mat([
          [-3, 9, 9, 0],
          [0, 5, 5, 0],
          [0, 0, 0, 0]
        ])
      );
    });

    test('row reduces a matrix with non-diagonal pivot entries', () => {
      const A = mat([
        [1, 1, 1],
        [0, 0, 2],
        [0, 0, 1]
      ]);

      const aRef = rowEchelonForm(A);
      expect(aRef.toArray()).toStrictEqual([
        [1, 1, 1],
        [0, 0, 2],
        [0, 0, 0]
      ]);
    });

    test('does nothing to an identity matrix', () => {
      const I = eye(3);
      const ref = rowEchelonForm(I);
      expect(ref.equals(I)).toBe(true);
    });

    test('handles the degenerate case', () => {
      expect(rowEchelonForm(NumberMatrix.builder().empty()).toArray()).toStrictEqual([]);
    });
  });

  describe('reducedRowEchelonForm', () => {
    test('row reduces a "wide" matrix', () => {
      const A = mat([
        [1, 2, 3],
        [4, 5, 6]
      ]);
      const aRef = reducedRowEchelonForm(A);
      expect(aRef.toArray()).toStrictEqual([
        [1, 0, -1],
        [0, 1, 2]
      ]);
    });

    test('row reduces a "tall" matrix', () => {
      const A = mat([
        [0, 1],
        [0, 0],
        [5, 9]
      ]);
      const aRef = reducedRowEchelonForm(A);
      expect(aRef.toArray()).toStrictEqual([
        [1, 0],
        [0, 1],
        [0, 0]
      ]);
    });

    test('row reduces a matrix with non-independent rows', () => {
      const A = mat([
        [1, 2, 3],
        [1, 1, 1],
        [1, 1, 1]
      ]);

      const aRef = reducedRowEchelonForm(A);
      expect(aRef.toArray()).toStrictEqual([
        [1, 0, -1],
        [0, 1, 2],
        [0, 0, 0]
      ]);
    });

    test('row reduces a matrix with non-diagonal pivot entries', () => {
      const A = mat([
        [1, 1, 1],
        [0, 0, 2],
        [0, 0, 1]
      ]);

      const aRef = reducedRowEchelonForm(A);
      expect(aRef.toArray()).toStrictEqual([
        [1, 1, 0],
        [0, 0, 1],
        [0, 0, 0]
      ]);
    });

    test('does nothing to an identity matrix', () => {
      const I = eye(3);
      const ref = reducedRowEchelonForm(I);
      expect(ref.equals(I)).toBe(true);
    });

    test('handles the degenerate case', () => {
      expect(reducedRowEchelonForm(mat([])).toArray()).toStrictEqual([]);
    });
  });

  describe('inverse', () => {
    test('does not affect an identity matrix', () => {
      const I = eye(3);
      const iInv = inverse(I);
      expect(iInv).not.toBeUndefined;
      expect((iInv as Matrix<number>).equals(I)).toBe(true);
    });

    test('calculates the inverse of a square matrix', () => {
      const A = mat([
        [4, 7],
        [2, 6]
      ]);
      const expectedInverse = mat([
        [0.6, -0.7],
        [-0.2, 0.4]
      ]);
      const aInv = inverse(A) as Matrix<number>;
      expect(aInv).not.toBeUndefined;
      expect(aInv.equals(expectedInverse)).toBe(true);
    });

    test('rejects a non-square matrix', () => {
      const nonSquare = mat([[0, 1]]);
      expect(() => inverse(nonSquare)).toThrow();
    });

    test('returns undefined for a singular matrix', () => {
      const S = mat([
        [0, 0],
        [0, 0]
      ]);
      const sInv = inverse(S);
      expect(sInv).toBeUndefined;
    });
  });
});
