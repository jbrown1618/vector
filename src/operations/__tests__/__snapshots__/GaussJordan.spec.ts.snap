// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GaussJordan solve returns the same results as solveByGaussianElimination 1`] = `
{
  "solution": FloatVector {
    "_data": Float64Array [
      0.9136263277500446,
      0.346696713406094,
      0.3111557882953938,
      0.39975222268696264,
      -0.1643066328933712,
      0.03885561914864894,
      -0.6657811157870935,
      0.007257352835084505,
      0.28777790799551173,
      -1.0613360257097249,
      0.9776891059889999,
      -0.22741440800345275,
      0.13036176562117635,
      -0.7378049693857984,
      -0.16142731728327223,
      -0.24265911924001132,
      -1.0766507493074127,
      -0.628984646265577,
      2.100922398641603,
      0.9878363505843968,
    ],
  },
  "solutionType": "Unique",
}
`;

exports[`GaussJordan solveByGaussianElimination solves a random 20x20 system 1`] = `
{
  "solution": FloatVector {
    "_data": Float64Array [
      0.9136263277500446,
      0.346696713406094,
      0.3111557882953938,
      0.39975222268696264,
      -0.1643066328933712,
      0.03885561914864894,
      -0.6657811157870935,
      0.007257352835084505,
      0.28777790799551173,
      -1.0613360257097249,
      0.9776891059889999,
      -0.22741440800345275,
      0.13036176562117635,
      -0.7378049693857984,
      -0.16142731728327223,
      -0.24265911924001132,
      -1.0766507493074127,
      -0.628984646265577,
      2.100922398641603,
      0.9878363505843968,
    ],
  },
  "solutionType": "Unique",
}
`;
